### ISSUE TITLE
CI Broken on master

### ISSUE THREAD

**Issue by icexelloss** [ibis-project/ibis#2040]
We started to see CI failure recently, for example:
https://dev.azure.com/ibis-project/ibis/_build/results?buildId=1702

The builds fails in Spark/PySpark test with this exception:
```
__________________ ERROR at setup of test_array_slice[None-3] __________________
[gw0] linux -- Python 3.6.7 /opt/conda/envs/ibis-env/bin/python

    @pytest.fixture(scope='session')
    def client():
        from pyspark.sql import SparkSession
        import pyspark.sql.functions as F
    
>       session = SparkSession.builder.getOrCreate()

ibis/pyspark/tests/conftest.py:11: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/conda/envs/ibis-env/lib/python3.6/site-packages/pyspark/sql/session.py:173: in getOrCreate
    sc = SparkContext.getOrCreate(sparkConf)
/opt/conda/envs/ibis-env/lib/python3.6/site-packages/pyspark/context.py:367: in getOrCreate
    SparkContext(conf=conf or SparkConf())
/opt/conda/envs/ibis-env/lib/python3.6/site-packages/pyspark/context.py:133: in __init__
    SparkContext._ensure_initialized(self, gateway=gateway, conf=conf)
/opt/conda/envs/ibis-env/lib/python3.6/site-packages/pyspark/context.py:316: in _ensure_initialized
    SparkContext._gateway = gateway or launch_gateway(conf)
/opt/conda/envs/ibis-env/lib/python3.6/site-packages/pyspark/java_gateway.py:46: in launch_gateway
    return _launch_gateway(conf)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

conf = <pyspark.conf.SparkConf object at 0x7fe7eeb7a9e8>, insecure = False

    def _launch_gateway(conf=None, insecure=False):
        """
        launch jvm gateway
        :param conf: spark configuration passed to spark-submit
        :param insecure: True to create an insecure gateway; only for testing
        :return: a JVM gateway
        """
        if insecure and os.environ.get("SPARK_TESTING", "0") != "1":
            raise ValueError("creating insecure gateways is only for testing")
        if "PYSPARK_GATEWAY_PORT" in os.environ:
            gateway_port = int(os.environ["PYSPARK_GATEWAY_PORT"])
            gateway_secret = os.environ["PYSPARK_GATEWAY_SECRET"]
        else:
            SPARK_HOME = _find_spark_home()
            # Launch the Py4j gateway using Spark's run command so that we pick up the
            # proper classpath and settings from spark-env.sh
            on_windows = platform.system() == "Windows"
            script = "./bin/spark-submit.cmd" if on_windows else "./bin/spark-submit"
            command = [os.path.join(SPARK_HOME, script)]
            if conf:
                for k, v in conf.getAll():
                    command += ['--conf', '%s=%s' % (k, v)]
            submit_args = os.environ.get("PYSPARK_SUBMIT_ARGS", "pyspark-shell")
            if os.environ.get("SPARK_TESTING"):
                submit_args = ' '.join([
                    "--conf spark.ui.enabled=false",
                    submit_args
                ])
            command = command + shlex.split(submit_args)
    
            # Create a temporary directory where the gateway server should write the connection
            # information.
            conn_info_dir = tempfile.mkdtemp()
            try:
                fd, conn_info_file = tempfile.mkstemp(dir=conn_info_dir)
                os.close(fd)
                os.unlink(conn_info_file)
    
                env = dict(os.environ)
                env["_PYSPARK_DRIVER_CONN_INFO_PATH"] = conn_info_file
                if insecure:
                    env["_PYSPARK_CREATE_INSECURE_GATEWAY"] = "1"
    
                # Launch the Java gateway.
                # We open a pipe to stdin so that the Java gateway can die when the pipe is broken
                if not on_windows:
                    # Don't send ctrl-c / SIGINT to the Java gateway:
                    def preexec_func():
                        signal.signal(signal.SIGINT, signal.SIG_IGN)
                    proc = Popen(command, stdin=PIPE, preexec_fn=preexec_func, env=env)
                else:
                    # preexec_fn not supported on Windows
                    proc = Popen(command, stdin=PIPE, env=env)
    
                # Wait for the file to appear, or for the process to exit, whichever happens first.
                while not proc.poll() and not os.path.isfile(conn_info_file):
                    time.sleep(0.1)
    
                if not os.path.isfile(conn_info_file):
>                   raise Exception("Java gateway process exited before sending its port number")
E                   Exception: Java gateway process exited before sending its port number

/opt/conda/envs/ibis-env/lib/python3.6/site-packages/pyspark/java_gateway.py:108: Exception
```

**Comment by icexelloss** [comment]
I am looking into this. I can reduce the CI failure by running the spark tests with docker-compose:

```
docker-compose -f ci/docker-compose.yml run ibis pytest -n auto -m 'pyspark'
```

However, it works when I run tests outside docker container
```
pytest -n auto -m 'pyspark'
```

**Comment by icexelloss** [comment]
Creating SparkSession directly (without pytest) in container also works:

```
docker-compose -f ci/docker-compose.yml run ibis python -c "from pyspark.sql import SparkSession; spark = SparkSession.builder.getOrCreate(); spark.range(10).show()"
```

**Comment by xmnlab** [comment]
@icexelloss I will also take a look into that.

**Comment by icexelloss** [comment]
Thanks! @xmnlab 

Update:

I have ping down the issue to be this line in pyspark:
https://github.com/apache/spark/blob/v2.4.4/python/pyspark/java_gateway.py#L98

It appears that the subprocess exited with 127 and the stderr of the subprocess is:
"/usr/bin/env: ‘bash’: No such file or directory"

The subprocess command is:
['/opt/conda/envs/ibis-env/lib/python3.6/site-packages/pyspark/./bin/spark-submit', 'pyspark-shell']

I cannot reproduce this error by hitting this code without pytest, i.e., start Pyspark directly in a python script:
```
from pyspark.sql import SparkSession
spark = SparkSession.builder.getOrCreate()
```

**Comment by icexelloss** [comment]
I have traced the issue to be related to the env gets used when call Popen:
When running with pytest, the `PATH` has an additional value, somehow this is causing Popen to fail. 

Here is a min reproduce:

```
import subprocess

# bad_env is the env when calling with pytest
bad_env={'PATH': '/opt/conda/envs/ibis-env/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin;/opt/conda/envs/ibis-env/lib'}

# good_env is the env when calling with python or pyspark
good_env={'PATH': '/opt/conda/envs/ibis-env/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'}

print("Test good env")
good_proc = subprocess.Popen(['/ibis/test.sh'], env=good_env)
print("Done: ", good_proc.wait())

print("Test bad env")
bad_proc = subprocess.Popen(['/ibis/test.sh'], env=bad_env)
print("Done: ", bad_proc.wait())
```

```
(ibis-env) root@eba7cd3f948f:/ibis# python test_shell.py
Test good env
Hello
Done:  0
Test bad env
/usr/bin/env: 'bash': No such file or directory
Done:  127
```

**Comment by icexelloss** [comment]
Actually, there is a `;` in the `PATH` when calling with pytest 5.2.4

**Comment by icexelloss** [comment]
Verified `;` is not a typo by me, but some how added by pytest:

```
(ibis-env) root@eba7cd3f948f:/ibis# pytest ibis/pyspark/tests/pytest_spark.py 
==================================================================== test session starts =====================================================================
platform linux -- Python 3.6.7, pytest-5.2.4, py-1.8.0, pluggy-0.12.0
rootdir: /ibis, inifile: setup.cfg
plugins: xdist-1.30.0, cov-2.8.1, forked-1.1.2
collected 1 item                                                                                                                                             

ibis/pyspark/tests/pytest_spark.py 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /ibis/ibis/pyspark/tests/pytest_spark.py(5)test_spark()
-> spark = SparkSession.builder.getOrCreate()
(Pdb) ll
  3  	def test_spark():
  4  	    import pdb; pdb.set_trace()
  5  ->	    spark = SparkSession.builder.getOrCreate()
  6  	    spark.range(10).show()
(Pdb) import os
(Pdb) os.environ['PATH']
'/opt/conda/envs/ibis-env/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin;/opt/conda/envs/ibis-env/lib'
(Pdb) (ibis-env) root@eba7cd3f948f:/ibis# pytest ibis/pyspark/tests/pytest_spark.py 
==================================================================== test session starts =====================================================================
platform linux -- Python 3.6.7, pytest-5.2.4, py-1.8.0, pluggy-0.12.0
rootdir: /ibis, inifile: setup.cfg
plugins: xdist-1.30.0, cov-2.8.1, forked-1.1.2
collected 1 item                                                                                                                                             

ibis/pyspark/tests/pytest_spark.py 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /ibis/ibis/pyspark/tests/pytest_spark.py(5)test_spark()
-> spark = SparkSession.builder.getOrCreate()
(Pdb) ll
  3  	def test_spark():
  4  	    import pdb; pdb.set_trace()
  5  ->	    spark = SparkSession.builder.getOrCreate()
  6  	    spark.range(10).show()
(Pdb) import os
(Pdb) os.environ['PATH']
'/opt/conda/envs/ibis-env/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin;/opt/conda/envs/ibis-env/lib'
(Pdb)
```

**Comment by xmnlab** [comment]
oh I see .. thanks for the updates .. I checked it out on pytest github issues .. I didn't find anything directly related to that ... maybe we can just pin the latest version of pytest or a previous version .. 

also it seems there is an issue on linuxbuilddocs ... not sure yet what is the problem ...

**Comment by xmnlab** [comment]
I run this test file inside the root folder .. and it worked ... so maybe it is related to something related to `conftest` ... not sure yet .. I am playing with that

**Comment by xmnlab** [comment]
it seems the problem could be related to ibis/__init__.py .. 
if I remove ibis/__init__.py file .. the new ibis/test_pyspark.py worked ...

**Comment by icexelloss** [comment]
geopandas is the culprit!

```
(ibis-env) root@a14be6873e30:/ibis# python
Python 3.6.7 | packaged by conda-forge | (default, Nov  6 2019, 16:19:42) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ['PATH']
'/opt/conda/envs/ibis-env/bin:/opt/conda/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
>>> import geopandas
>>> os.environ['PATH']
'/opt/conda/envs/ibis-env/bin:/opt/conda/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin;/opt/conda/envs/ibis-env/lib'
vvv

**Comment by icexelloss** [comment]
Update: The actual culprit is rtree, filed https://github.com/Toblerity/rtree/issues/126 with rtree devs.

I am going to try to ping rtree version to see if ci works

**Comment by xmnlab** [comment]
great news! I hope it fix the issue :) thanks @icexelloss for work on that!

**Comment by jreback** [comment]
closed by #2043
